### DISCUSSION TITLE
Are channels closed by garbage collector?

### DISCUSSION THREAD

**Discussion by amano-kenji** [janet-lang/janet#1379] [Q&A] [answered]
I don't know whether garbage collector closes channels.

**Comment by CosmicToast** [comment]
When a channel gets garbage-collected, it [calls](https://github.com/janet-lang/janet/blob/cc5beda/src/core/ev.c#L1151) the [deinit](https://github.com/janet-lang/janet/blob/cc5beda/src/core/ev.c#L681) function.
Closing a channel calls [cfun_channel_close](https://github.com/janet-lang/janet/blob/cc5beda/src/core/ev.c#L1148).

These aren't exactly the same,
closing the channel makes all queued and future reads and writes return nil,
while deinitializing deinitializes (a simple [free](https://github.com/janet-lang/janet/blob/cc5beda0d26c0b91fb28540291e950392cf59fc4/src/core/ev.c#L123)) both of those queues.

Whether this makes any difference is a question for your code.
Test it out!

If you want to make it such that the answer is an unambiguous "yes", just add a constructed call of cfun_channel_close to the gc function before deinitializing.

**Reply by amano-kenji** [reply]
I think garbage collection doesn't kick in for a channel as long as the channel is used.

Garbage collection kicks in only after the channel stops being referenced anywhere.

**Reply by amano-kenji** [reply, marked as answer]
I think the current codebase does the right thing.
